xtm/xtm_scsp_queue.h

```diff
@@ -0,0 +1,161 @@
+#pragma once
+/*
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+template <class T>
+struct xtm_scsp_queue_read_iterator;
+
+/**
+ * Lock free single consumer, single producer queue,
+ * based on ring buffer.
+ * Template parameter can be of any type for which the
+ * copying operation is carried out quickly and correctly.
+ */
+template<class T>
+struct xtm_scsp_queue {
+	friend struct xtm_scsp_queue_read_iterator<T>;
+	/**
+	 * Init scsp queue struct.
+	 * @param[in] size - size of queue ring buffer, must be power of two,
+	 *                   for fast calculation of read/write position, using
+	 *                   '& (size - 1' operation.
+	 * @retval 0 if success, otherwise return -1.
+	 */
+	int
+	create(unsigned size) {
```

Comment:
Shouldn't we use Tarantool C style guide here (yes, I see that this is not C)? Type on the same line (looks better in grep), curly braces for functions on the next line.

Reply:
@Gerold103 what you prefer here?

Reply:
I would prefer C-like style. All like void234 said. But we don't have C++ guidelines anyway, so I can't force them.

Reply:
Fixed according to tarantool C code style